spectator/index.js

```diff
@@ -0,0 +1,193 @@
+                this.x = obj.x;
+                this.y = obj.y;
+                this.angle = obj.angle;
+        }
+
+        move(x, y) {
+                this.x = x;
+                this.y = y;
+        }
+
+        rotate(theta) {
+                this.theta = theta;
+        }
+
+        draw(ctx) {
+                ctx.save()
+                ctx.translate(this.x, this.y);
+
+                let oldFill = ctx.fillStyle;
+                ctx.beginPath();
+                ctx.arc(0, 0, 2, 0, 2 * Math.PI);
+                ctx.fillStyle = "#e05d5d";
+                ctx.fill();
+                ctx.fillStyle = oldFill;
+
+                ctx.restore();
+        }
+}
+
+var last_drawn_scoreboard = {};
+var initCanvas = false;
+connect(function (json) {
+        if (json.e === "teamnames") {
+                team_names = json.data;
+        } else if (json.e === "state") {
+                const data = json.data;
+                ctx.save()
+                ctx.clearRect(0, 0, c.width, c.height);
+                ctx.strokeStyle = "#ffffff";
+                ctx.lineWidth = 1;
+                ctx.lineCap = "square";
+                ctx.lineJoin = "bevel";
+
+                scaleXRatio = c.width / data.bounds[0];
+                scaleYRatio = c.height / data.bounds[0];
+                scaleRatio = Math.min(scaleXRatio, scaleYRatio);
+                ctx.transform(scaleRatio, 0, 0, scaleRatio, 0, 0);
+
+                ctx.beginPath();
+                ctx.moveTo(0, 0);
+                ctx.lineTo(data.bounds[0], 0);
+                ctx.lineTo(data.bounds[0], data.bounds[1]);
+                ctx.lineTo(0, data.bounds[1]);
+                ctx.lineTo(0, 0);
+                ctx.stroke();
+
+                for (const player of data.players) {
+                        new Ship(player).draw(ctx);
+                }
+
+                for (const bullet of data.bullets) {
+                        new Bullet(bullet).draw(ctx);
+                }
+                ctx.restore()
+
+                if (JSON.stringify(data.scoreboard) !== JSON.stringify(last_drawn_scoreboard)) {
+                        draw_scoreboard(data.scoreboard);
+                        last_drawn_scoreboard = data.scoreboard;
+                }
+        }
+});
+
+function draw_scoreboard(scoreboard) {
+        var sorted_players = Object.keys(scoreboard).sort(function (a, b) { return scoreboard[b] - scoreboard[a] });
+        chart.innerHTML = "";
+        for (let i = 0; i < sorted_players.length; i++) {
+                const player_id = sorted_players[i];
+                const player_score = String(scoreboard[player_id]).padEnd(3);
+                const team_name = team_names[player_id];
+                chart.innerHTML += `
+            <tr class="rank-${i + 1}">
+              <td class="rank">${i + 1}</span>
+              <td class="name">${team_name}</span>
+              <td class="score">${player_score}</span>
+            </tr>`;
+        }
+}
```

Comment:
EOL plz :))

Reply:
bloody hell hang on. I'll fix this real quick

Reply:
oh wait it's already there :smile: